redbpf/src/ringbuf.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2019 Authors of Red Sift
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! # Ring Buffer Event handling
+//!
+//! The ringbuf module makes it easier to hook up and consume from ring buffers.
+//! and provdes a safe interface for accesing the ring buffer
+//!
+//! The consumed data is a raw pointer and will require unsafe code to transform
+//! into a data structure.
+
+use crate::{Error, Map, Result};
+use std::cell::RefCell;
+use std::io;
+use std::os::unix::io::RawFd;
+use std::slice;
+
+use bpf_sys::{
+    ring_buffer, ring_buffer__consume, ring_buffer__epoll_fd, ring_buffer__free, ring_buffer__new,
+};
+use libc::c_void;
+
+#[derive(Debug)]
+pub struct RingBufMapContext {
+    messages: RefCell<Vec<Box<[u8]>>>,
+}
+
+impl RingBufMapContext {
+    fn new() -> RingBufMapContext {
+        RingBufMapContext {
+            messages: RefCell::new(Vec::new()),
+        }
+    }
+
+    pub fn read_message(&self, data: *mut c_void, size: u64) {
+        unsafe {
+            let mut messages = self.messages.borrow_mut();
+            let message = slice::from_raw_parts(data as *const u8, size as usize).to_vec();
+            messages.push(message.into_boxed_slice());
+        }
+    }
+
+    pub fn take_messages(&self) -> Vec<Box<[u8]>> {
+        let mut messages = self.messages.borrow_mut();
+        let new_vec = messages.clone();
+        messages.clear();
+        new_vec
+    }
+}
+
+#[derive(Debug)]
+pub struct RingBufInner(*mut ring_buffer);
+unsafe impl Send for RingBufInner {}
```

Comment:
I wonder this is required by `tokio`?

Reply:
Yes, that's right:
```
error[E0277]: `*mut ring_buffer` cannot be sent between threads safely
   --> /home/vagrant/redbpf/redbpf/src/load/loader.rs:73:13
    |
73  |             tokio::spawn(fut);
    |             ^^^^^^^^^^^^ `*mut ring_buffer` cannot be sent between threads safely
    |
   ::: /home/vagrant/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.15.0/src/task/spawn.rs:127:21
    |
127 |         T: Future + Send + 'static,
    |                     ---- required by this bound in `tokio::spawn`
    |
    = help: within `ForEach<RingBufMessageStream, futures::future::Ready<()>, [closure@/home/vagrant/redbpf/redbpf/src/load/loader.rs:69:39: 72:14]>`, the trait `std::marker::Send` is not implemented for `*mut ring_buffer`
note: required because it appears within the type `RingBufInner`
   --> /home/vagrant/redbpf/redbpf/src/ringbuf.rs:56:12
    |
56  | pub struct RingBufInner(*mut ring_buffer);
    |            ^^^^^^^^^^^^
note: required because it appears within the type `ringbuf::RingBufMap`
   --> /home/vagrant/redbpf/redbpf/src/ringbuf.rs:59:12
    |
59  | pub struct RingBufMap {
    |            ^^^^^^^^^^
note: required because it appears within the type `RingBufMessageStream`
   --> /home/vagrant/redbpf/redbpf/src/load/map_io.rs:73:12
    |
73  | pub struct RingBufMessageStream {
    |            ^^^^^^^^^^^^^^^^^^^^
    = note: required because it appears within the type `ForEach<RingBufMessageStream, futures::future::Ready<()>, [closure@/home/vagrant/redbpf/redbpf/src/load/loader.rs:69:39: 72:14]>`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `redbpf` due to previous error
error: failed to compile the `pidtree' program
```